flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContextTest.java

```diff
@@ -155,6 +156,41 @@ private void testAssignSplitToUnregisterdReader(boolean fromCoordinatorExecutor)
                 "Cannot assign splits");
     }
 
+    @Test
+    public void testExceptionInRunnableFailsTheJob() throws InterruptedException {
+        ManuallyTriggeredScheduledExecutorService manualWorkerExecutor =
+                new ManuallyTriggeredScheduledExecutorService();
+        // do not use ManuallyTriggeredScheduledExecutorService for coordinatorExecutor
+        // as it needs to use factory.newThread() so that the FailJobExceptionHandler is set
+        // use a custom ThreadPool to gain access to the underlying thread
+        // to wait for it before asserting
+        final ThreadAccessibleThreadPool coordinatorExecutor =
+                new ThreadAccessibleThreadPool(coordinatorThreadFactory);
+        SourceCoordinatorContext<MockSourceSplit> testingContext =
+                new SourceCoordinatorContext<>(
+                        coordinatorExecutor,
+                        manualWorkerExecutor,
+                        coordinatorThreadFactory,
+                        operatorCoordinatorContext,
+                        new MockSourceSplitSerializer(),
+                        splitSplitAssignmentTracker);
+
+        testingContext.runInCoordinatorThread(
+                () -> {
+                    throw new RuntimeException();
+                });
+
+        manualWorkerExecutor.triggerAll();
+        // testingContext.close(); already shutdowns coordinatorExecutor and blocks until tasks are
+        // finished
+        testingContext.close();
+        // wait that the thread handles the exception through the provided exception handler.
+        // Using ThreadPoolExecutor#awaitTermination is not enough to wait for exception handling.
+        // It just waits for the runnable
+        coordinatorExecutor.getSingleThread().join();
```

Comment:
I agree it is simpler and higher level. I did not do that because I did not want to change the signature of the existing test components but fair enough, I'll do that. Thanks for the suggestion.

Reply:
@tillrohrmann just a confirmation, you want `MockOperatorCoordinatorContext#failjob()` to become asynchronous? That would require passing the `ExecutorService` to `MockOperatorCoordinatorContext` and blocking on the Future termination in many tests. Maybe I misunderstood ?

Reply:
Makes total sense, thanks Till !

Reply:
That being said, master as moved `coordinatorExecutor` to be `ScheduledExecutorService` which catches the exceptions in `ScheduledFutureTask` so the exception handling works differently now. So I need to adapt the production code as well.